pam/internal/adapter/pwquality_c.go

```diff
@@ -0,0 +1,46 @@
+package adapter
+
+// #cgo pkg-config: pwquality
+// #include <stdio.h>
+// #include <stdlib.h>
+// #include <pwquality.h>
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"unsafe"
+)
+
+// checkChallengeQuality checks the quality of the new password using the pwquality library.
+func checkChallengeQuality(old, new string) error {
+	pwq := C.pwquality_default_settings()
+	if pwq == nil {
+		return errors.New("could not allocate pw quality default settings")
+	}
+	defer C.pwquality_free_settings(pwq)
+
+	var auxErr *C.char
+	auxErrPointer := unsafe.Pointer(auxErr)
+	defer C.free(auxErrPointer)
+
+	// Load pwquality configuration (from /etc/security/pwquality.conf)
+	if ret := C.pwquality_read_config(pwq, nil, &auxErrPointer); ret < 0 {
+		var buf [C.PWQ_MAX_ERROR_MESSAGE_LEN]C.char
+		errMsg := C.GoString(C.pwquality_strerror(&buf[0], C.size_t(len(buf)), ret, auxErrPointer))
+		return fmt.Errorf("can't ready pwquality configuration: %s", errMsg)
+	}
+
+	oldC := C.CString(old)
+	defer C.free(unsafe.Pointer(oldC))
+
+	newC := C.CString(new)
+	defer C.free(unsafe.Pointer(newC))
+
+	if ret := C.pwquality_check(pwq, newC, oldC, nil, &auxErrPointer); ret < 0 {
+		var buf [C.PWQ_MAX_ERROR_MESSAGE_LEN]C.char
+		errMsg := C.GoString(C.pwquality_strerror(&buf[0], C.size_t(len(buf)), ret, auxErrPointer))
+		return fmt.Errorf("password quality failed: %s", errMsg)
```

Comment:
_password quality **check** failed_ ?

However, given that the return value of C.pwquality_strerror is something meant to be used by the UI (being translated and eveything), shouldn't we return it as an extra value so that we can use it as it is as our error message?

Ideally it can be just returned as a different error type so that we just return one single error, that eventually we can parse via `errors.As`.

Reply:
This error is only used in a very specific context (and it doesn't care if it's a specific type or not, since the context of when it happens only deals in string messages anyway), so I think it's a bit of overengineering to handle all of this, no?

Reply:
I mean, since we've a prefabricated error message to show to the UI, maybe we can just show that without our not-translated (and in go-error style) prefix.

So to make it easier, we can just return `errros.New(errMsg)` here if you want to keep it simple

Reply:
Saw it changed, good... need the other too though :)